src/refiners/foundationals/latent_diffusion/schedulers/euler.py

```diff
@@ -22,31 +24,45 @@ def __init__(
             device=device,
             dtype=dtype,
         )
-        self.sigmas = self.noise_std / self.cumulative_scale_factors
-        self.sigmas[0] = 0.0
+        self.sigmas = self._generate_sigmas().to(device=self.device, dtype=self.dtype)
 
     def _generate_timesteps(self) -> Tensor:
+        # using "trailing" timestep
         step_ratio = self.num_train_timesteps / self.num_inference_steps
         timesteps = arange(1000, 0, -step_ratio).round().type(int64)
         return timesteps - 1
 
-    def __call__(self, x: Tensor, noise: Tensor, step: int) -> Tensor:
-        timestep, previous_timestep = (
-            self.timesteps[step],
-            (
-                self.timesteps[step + 1]
-                if step < self.num_inference_steps - 1
-                else tensor(data=[0], device=self.device, dtype=self.dtype)
-            ),
-        )
-        current_sigma, previous_sigma = self.sigmas[timestep], (
-            self.sigmas[previous_timestep] if previous_timestep > 0 else self.sigmas[0]
+    def _generate_sigmas(self) -> Tensor:
+        sigmas = self.noise_std / self.cumulative_scale_factors
+        sigmas = torch.from_numpy(
+            np.interp(self.timesteps.cpu().numpy(), np.arange(0, len(sigmas)), sigmas.cpu().numpy())
         )
+        sigmas = torch.cat([sigmas, tensor([0.0])])
+        print(sigmas)
+        return sigmas
 
-        predicted_x = x - current_sigma * noise
+    def __call__(
+        self,
+        x: Tensor,
+        noise: Tensor,
+        step: int,
+        s_churn: float = 0.0,
+        s_tmin: float = 0.0,
+        s_tmax: float = float("inf"),
+        s_noise: float = 1.0,
+    ) -> Tensor:
+        sigma = self.sigmas[step]
 
-        derivative = (x - predicted_x) / current_sigma
-        dt = previous_sigma - current_sigma
-        denoised_x = x + derivative * dt
+        gamma = min(s_churn / (len(self.sigmas) - 1), 2**0.5 - 1) if s_tmin <= sigma <= s_tmax else 0
 
+        eps = noise * s_noise
+        sigma_hat = sigma * (gamma + 1)
+        if gamma > 0:
+            x = x + eps * (sigma_hat**2 - sigma**2) ** 0.5
+
+        predicted_x = x * torch.sqrt(sigma_hat**2 + 1) - eps * sigma_hat
```

Comment:
Not sure to understand this one. E.g. diffusers reference [implementation](https://github.com/huggingface/diffusers/blob/08b453e3828f80027d881bb460716af95e192bcd/src/diffusers/schedulers/scheduling_euler_discrete.py#L496) would translate into:

```python
predicted_x = x - sigma_hat * noise
```

Reply:
In the previous commit, I was using the one you find [here](https://github.com/finegrain-ai/refiners/pull/138#issuecomment-1845368366). It generates brown images similar to those in the provided link. In the paper, they refer to it as $D(x;t)$, so I utilized the predicted $x_0$ from the DDIM implementation adapted to be computed by sigma_hat. That's how I obtained somewhat better samples. I believe the issue may be related to scaling.